crates/libs/windows/src/core/literals.rs

```diff
@@ -0,0 +1,147 @@
+/// Simply terminates the literal UTF-8 string.
+#[macro_export]
+macro_rules! s {
+    ($s:literal) => {
+        ::windows::core::PCSTR(::std::concat!($s, '\0').as_ptr())
+    };
+}
+
+/// Converts the literal UTF-8 string into a UTF-16 string adding a terminator and then wrapping
+/// that in an HSTRING reference so that it can be used for calling both WinRT APIs expecting an
+/// HSTRING as well as Win32 APIs expecting a PCWSTR.
+#[macro_export]
+macro_rules! w {
+    ($s:literal) => {{
+        const INPUT: &[u8] = $s.as_bytes();
+        const OUTPUT_LEN: usize = ::windows::core::utf16_len(INPUT) + 1;
+        const RESULT: ::windows::core::HSTRING = {
+            if OUTPUT_LEN == 1 {
+                unsafe { ::std::mem::transmute(::std::ptr::null::<u16>()) }
+            } else {
+                const OUTPUT: &[u16; OUTPUT_LEN] = {
+                    let mut buffer = [0; OUTPUT_LEN];
+                    let mut input_pos = 0;
+                    let mut output_pos = 0;
+                    while let Some((mut code_point, new_pos)) = ::windows::core::decode_utf8(INPUT, input_pos) {
+                        input_pos = new_pos;
+                        if code_point <= 0xffff {
+                            buffer[output_pos] = code_point as u16;
+                            output_pos += 1;
+                        } else {
+                            code_point -= 0x10000;
+                            buffer[output_pos] = 0xd800 + (code_point >> 10) as u16;
+                            output_pos += 1;
+                            buffer[output_pos] = 0xdc00 + (code_point & 0x3ff) as u16;
+                            output_pos += 1;
+                        }
+                    }
+                    &{ buffer }
+                };
+                const HEADER: ::windows::core::HSTRING_HEADER = ::windows::core::HSTRING_HEADER { flags: 0x11, len: (OUTPUT_LEN - 1) as u32, padding1: 0, padding2: 0, ptr: OUTPUT.as_ptr() };
+                unsafe { ::std::mem::transmute(&HEADER) }
```

Comment:
```suggestion
               // SAFETY: an `HSTRING` is exactly equivalent to a pointer to an `HSTRING_HEADER`
                unsafe { ::std::mem::transmute::<&::windows::core::HSTRING_HEADER, ::windows::core::HSTRING>(&HEADER) }
```

Using explicit types when using `transmute` is a safer option. This way, there's no risk of code else where in this `const` changing and then changing the type inference in unexpected ways.